### DISCUSSION TITLE
The multi value assets query param order affects the output of the `stac/tiles` endpoint.

### DISCUSSION THREAD

**Discussion by raster-blaster** [developmentseed/titiler#645] [Q&A] [answered]
I have noticed that the order of the `assets` query parameter affects the output of the `stac/tiles` endpoint in `Titiler`. For example, the following two requests have the same parameters, except for the order of the `assets` parameter, but they produce different outputs:

* https://api.cogeo.xyz/stac/tiles/WebMercatorQuad/13/4794/3394@1x?url=https%3A%2F%2Fearth-search.aws.element84.com%2Fv0%2Fcollections%2Fsentinel-s2-l2a-cogs%2Fitems%2FS2B_36RTT_20190102_0_L2A&resampling_method=nearest&assets=B04%2CB03%2CB02&bidx=1&rescale=0%2C1000
*  https://api.cogeo.xyz/stac/tiles/WebMercatorQuad/13/4794/3394@1x?url=https%3A%2F%2Fearth-search.aws.element84.com%2Fv0%2Fcollections%2Fsentinel-s2-l2a-cogs%2Fitems%2FS2B_36RTT_20190102_0_L2A&resampling_method=nearest&assets=B02%2CB03%2CB04&bidx=1&rescale=0%2C1000

In the `rio-tiler` utility render function, the provided assets are stacked as a 3-dimensional NumPy array and passed to `rasterio` to create an RGB image with the [1,2,3] color indexes. The assets are stacked in the same order as we provide in the `stac/tiles` endpoint.

Now, I have a few questions:

1. Is preserving the order of the `assets` parameter the expected behavior in `Titiler`?
2. Since HTTP requests do not guarantee the order of query parameters or array query parameters, isn't this problematic? If the order of the assets parameter changes on the fly, different color combinations will be produced, resulting in images with different channel arrangements.
3. Is there a way to specify the order of indexes or which `asset` corresponds to a particular color channel in the `stac/tiles` endpoint?
4. Does `Titiler`, underneath `FastAPI`, preserve the order of the `assets` query parameter?
5. Is it the responsibility of the client to ensure the correct order of the `assets` query parameter in the `stac/tiles` endpoint?

**Comment by vincentsarago** [comment, marked as answer]
Hi @raster-blaster 

Multiple things:
- api.cogeo.xyz is deprecated and use a really old version of titiler 
- `https://titiler.xyz` is a official demo endpoint 
- in your case passing `assets=B04,B03,B02` and `assets=B02,B03,B04` is expected to gives different result, titiler will respect the order it was given

> Since HTTP requests do not guarantee the order of query parameters or array query parameters, isn't this problematic?

Is that true? if yes this is problematic and we have then a bug in titiler, but TBH for now I never had any problem with parameter order 🤷 

> Does Titiler, underneath FastAPI, preserve the order of the assets query parameter? 

It should

> Is it the responsibility of the client to ensure the correct order of the assets query parameter in the stac/tiles endpoint?

Yes, TiTiler won't decide things for users 😄
